### PR TITLE
MC-85: Add category processor, normalizer and dictionary

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | y
| Specs pass?*         | y
| Checkstyle issues?** | n
| PMD issues?***       | n
| Fixed tickets        | MC-85
| Doc PR               |

### CODE REVIEW

File: Normalizer/CategoryNormalizer.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    {
+        $this->categoryManager = $categoryManager;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($object, $format = null, array $context = [])
+    {
+        $normalized    = [];
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: useless

File: Normalizer/CategoryNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        $defaultLocale = $context['defaultLocale'];
+        $categoryTree  = $this->getTranslatedCategoryTree($object, $defaultLocale);
+        $root          = array_shift($categoryTree);
+        $rootName      = $context['userCategoryMapping'][$root];
+        $categoryPath  = implode('/', $categoryTree);
+        $categoryName  = $this->getTranslatedName($object, $defaultLocale);
+
+        $normalized[] = [
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: $normalized

File: Normalizer/CategoryNormalizer.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        ];
+
+        $updateParts = $this->getTheStoreViewsUpdateParts(
+            $object,
+            $context['storeViewMapping'],
+            $categoryPath,
+            $rootName
+        );
+
+        return array_merge($normalized, $updateParts);
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: [$normalized]

File: Normalizer/CategoryNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+     *
+     * @param CategoryInterface $category
+     * @param array             $storeViewMapping
+     * @param string            $rootName
+     *
+     * @returns array
+     */
+    protected function getTheStoreViewsUpdateParts(CategoryInterface $category, array $storeViewMapping, $rootName)
+    {
+        $updateParts = [];
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: Useless

File: Normalizer/CategoryNormalizer.php [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+{
+    /** @var CategoryManager */
+    protected $categoryManager;
+
+    /**
+     * @param CategoryManager $categoryManager
+     */
+    public function __construct(CategoryManager $categoryManager)
+    {
+        $this->categoryManager = $categoryManager;
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: Why don't you directly inject the CategoryRepository?

File: Normalizer/CategoryNormalizer.php (hunk 4 above)
Comment (willy-ahva) [inline Normalizer/CategoryNormalizer.php]: Nope, I need to return an array if there is no mapping.

Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: You're right ^^

File: Normalizer/CategoryNormalizer.php [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+     *
+     * @param array $fullCatTree
+     *
+     * @return string
+     */
+    protected function getCategoryPath(array $fullCatTree)
+    {
+        array_shift($fullCatTree);
+
+        return implode('/', $fullCatTree);
```
Comment (fitn) [inline Normalizer/CategoryNormalizer.php]: There is no existing constant for that? Can you look at NestedCategoryTree in gedmo extension?

File: Resources/config/processors.yml [hunk 7]
```diff
@@ -5,6 +5,7 @@ parameters:
     pim_magento_connector.processor.attribute.class: Pim\Bundle\MagentoConnectorBundle\Processor\AttributeProcessor
     pim_magento_connector.processor.family.class: Pim\Bundle\MagentoConnectorBundle\Processor\FamilyProcessor
     pim_magento_connector.processor.add_attribute_to_sets.class: Pim\Bundle\MagentoConnectorBundle\Processor\AddAttributeToSetsProcessor
+    pim_magento_connector.processor.category.class: Pim\Bundle\MagentoConnectorBundle\Processor\CategoryToArrayProcessor
```
Comment (BitOne) [inline Resources/config/processors.yml]: A nice alignment of the parameter value could help readability.

File: Resources/config/serializers.yml [hunk 8]
```diff
@@ -6,6 +6,7 @@ parameters:
     pim_magento_connector.normalizer.attribute_option.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\AttributeOptionNormalizer
     pim_magento_connector.normalizer.metric.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\MetricNormalizer
     pim_magento_connector.normalizer.datetime.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\DateTimeNormalizer
+    pim_magento_connector.normalizer.category.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\CategoryNormalizer
     pim_magento_connector.normalizer.collection.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\CollectionNormalizer
     pim_magento_connector.normalizer.product_category.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\ProductCategoryNormalizer
     pim_magento_connector.normalizer.product_media.class: Pim\Bundle\MagentoConnectorBundle\Normalizer\ProductMediaNormalizer
```
Comment (BitOne) [inline Resources/config/serializers.yml]: Same here about the alignement of parameters. It's the norm on the PIM, so let's follow that.

File: Resources/config/processors.yml (hunk 7 above)
Comment (willy-ahva) [inline Resources/config/processors.yml]: That means if I add a new longer param, all lines will be impacted and we will see modifications on github while there is no real modifs. I not really agree with it.

File: Normalizer/CategoryNormalizer.php [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function normalize($category, $format = null, array $context = [])
+    {
+        $defaultLocale = $context['defaultLocale'];
+        $rootName      = $this->getRootName($category, $context['userCategoryMapping'], $defaultLocale);
+        $categoryPath  = $this->getCategoryPath($category, $defaultLocale);
```
Comment (BitOne) [inline Normalizer/CategoryNormalizer.php]: getFormattedPath would be more explicit ?

File: Normalizer/CategoryNormalizer.php [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+
+        foreach ($categoryPath as $categoryNode) {
+            $translatedPath[] = $this->getTranslatedName($categoryNode, $locale);
+        }
+
+        return $translatedPath;
+    }
+
+    /**
+     * Returns the category path imploding with / and without root
```
Comment (BitOne) [inline Normalizer/CategoryNormalizer.php]: Good description :+1:

Comment (willy-ahva) [inline Normalizer/CategoryNormalizer.php]: Don't hesistate if you see bad description or bad return description. I hated it when I was young in Akeneo :p

File: Normalizer/CategoryNormalizer.php [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+     *
+     * @param CategoryInterface $category
+     * @param array             $storeViewMapping
+     * @param string            $rootName
+     *
+     * @returns array
+     */
+    protected function getStoreViewParts(CategoryInterface $category, array $storeViewMapping, $rootName)
+    {
+        $updateParts = [];
```
Comment (BitOne) [inline Normalizer/CategoryNormalizer.php]: So it's never used when you create a category for the first time ?
Or maybe it should be called storeviewParts ?

File: Normalizer/CategoryNormalizer.php [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    protected function getFormattedPath(CategoryInterface $category, $locale)
+    {
+        $translatedPath = [];
+        $categoryPath = $this->categoryRepository->getPath($category);
+
+        foreach ($categoryPath as $categoryNode) {
+            $translatedPath[] = $this->getTranslatedName($categoryNode, $locale);
+        }
+
+        unset($translatedPath[0]);
```
Comment (BitOne) [inline Normalizer/CategoryNormalizer.php]: I would have do the unset before the foreach, on the categoryPath.

**Pull request merged**